src/shell-util.c

```diff
@@ -492,3 +499,117 @@ shell_util_composite_capture_images (ClutterCapture  *captures,
+{
+  FindLatinKeysymsState *state = data;
+  int n_keysyms, i;
+  const xkb_keysym_t *keysyms;
+
+  n_keysyms = xkb_keymap_key_get_syms_by_level (keymap,
+                                                key,
+                                                0,
+                                                0,
+                                                &keysyms);
+  for (i = 0; i < n_keysyms; i++)
+    {
+      xkb_keysym_t keysym = keysyms[i];
+      if (keysym >= XKB_KEY_a && keysym <= XKB_KEY_z)
+        {
+          unsigned int keysym_index = keysym - XKB_KEY_a;
+          if (!state->required_keysyms_found[keysym_index])
+            {
+              state->required_keysyms_found[keysym_index] = TRUE;
+              state->n_required_keysyms--;
+            }
+        }
+    }
+}
+
+static struct xkb_keymap *
+create_keymap(struct xkb_context *context,
+              const char         *layouts,
+              const char         *variants,
+              const char         *options)
+{
+    struct xkb_keymap *keymap;
+    struct xkb_rule_names rmlvo = {
+        .rules = "evdev",
+        .model = "pc104",
+        .layout = layouts,
+	.variant = variants,
+        .options = options
+    };
+
+    if (!layouts && !variants && !options)
+        keymap = xkb_keymap_new_from_names (context, NULL, 0);
+    else
+        keymap = xkb_keymap_new_from_names (context, &rmlvo, 0);
+    if (!keymap)
+      {
+        fprintf (stderr,
+                 "Failed to compile RMLVO: '%s', '%s', '%s'\n",
+		 layouts, variants, options);
+        return NULL;
+      }
+
+    return keymap;
+}
+
+gboolean
+shell_util_needs_secondary_layout (const char *layouts,
+                                   const char *variants,
+                                   const char *options)
+{
+  gboolean required_keysyms_found[] = {
+    FALSE, /* XKB_KEY_a */
+    FALSE, /* XKB_KEY_b */
+    FALSE, /* XKB_KEY_c */
+    FALSE, /* XKB_KEY_d */
+    FALSE, /* XKB_KEY_e */
+    FALSE, /* XKB_KEY_f */
+    FALSE, /* XKB_KEY_g */
+    FALSE, /* XKB_KEY_h */
+    FALSE, /* XKB_KEY_i */
+    FALSE, /* XKB_KEY_j */
+    FALSE, /* XKB_KEY_k */
+    FALSE, /* XKB_KEY_l */
+    FALSE, /* XKB_KEY_m */
+    FALSE, /* XKB_KEY_n */
+    FALSE, /* XKB_KEY_o */
+    FALSE, /* XKB_KEY_p */
+    FALSE, /* XKB_KEY_q */
+    FALSE, /* XKB_KEY_r */
+    FALSE, /* XKB_KEY_s */
+    FALSE, /* XKB_KEY_t */
+    FALSE, /* XKB_KEY_u */
+    FALSE, /* XKB_KEY_v */
+    FALSE, /* XKB_KEY_w */
+    FALSE, /* XKB_KEY_x */
+    FALSE, /* XKB_KEY_y */
+    FALSE, /* XKB_KEY_z */
+  };
+  FindLatinKeysymsState state = {
+    .required_keysyms_found = required_keysyms_found,
+    .n_required_keysyms = G_N_ELEMENTS (required_keysyms_found),
+  };
+  struct xkb_keymap *keymap;
+  struct xkb_context *ctx;
+  enum xkb_context_flags ctx_flags;
+
+  ctx_flags = XKB_CONTEXT_NO_DEFAULT_INCLUDES;
+  ctx = xkb_context_new (ctx_flags);
+  xkb_context_include_path_append_default (ctx);
```

Comment:
If you adjust the ordering of the declarations above, you can declare and initialise these as the same time.

Reply:
(That is unless the `create_keymap` call has a dependency on `xkb_context_include_path_append_default`)

Reply:
(At which point my personal inclination would be to wrap both `xkb_context_new` and that ^ in another function and return the created context).